lib/Arch.cpp

```diff
@@ -0,0 +1,441 @@
+        anvill::ValueDecl value_declaration = {};
+
+        const llvm::Value *value = return_inst->getReturnValue();
+        if (!value) continue;
+        value_declaration.type = value->getType();
+
+        switch (value_declaration.type->getTypeID()) {
+          case llvm::Type::IntegerTyID:
+          case llvm::Type::PointerTyID: {
+            // Allocate RAX for an integer or pointer
+            value_declaration.reg =
+                new remill::Register("RAX", 0, 8, 0, value_declaration.type);
+            break;
+          }
+          case llvm::Type::FloatTyID: {
+            // Allocate XMM0 for a floating point value
+            value_declaration.reg =
+                new remill::Register("XMM0", 0, 16, 0, value_declaration.type);
+            break;
+          }
+          case llvm::Type::StructTyID: {
+            // Try to split the struct over the registers
+            std::vector<bool> allocated(return_register_constraints.size(),
+                                        false);
+            auto struct_ptr =
+                llvm::cast<llvm::StructType>(value_declaration.type);
+            auto mapping = TryReturnThroughRegisters(
+                *struct_ptr, return_register_constraints);
+            if (mapping) {
+              // There is a valid split over registers, so add the mapping
+              return *mapping;
+            } else {
+              // Struct splitting didn't work so do RVO. Assume that the pointer
+              // to the return value resides in RAX.
+              value_declaration.reg =
+                  new remill::Register("RAX", 0, 8, 0, value_declaration.type);
+            }
+            break;
+          }
+          default: {
+            LOG(ERROR) << "Encountered an unknown return type";
+            exit(1);
+          }
+        }
+
+        return_value_declarations.push_back(value_declaration);
+      }
+    }
+  }
+
+  return return_value_declarations;
+}
+
+std::vector<anvill::ParameterDecl> X86_64_SysV::BindParameters(
+    const llvm::Function &function) {
+  std::vector<anvill::ParameterDecl> parameter_declarations;
+  auto param_names = TryRecoverParamNames(function);
+  llvm::DataLayout dl(function.getParent());
+
+  // Used to keep track of which registers have been allocated
+  std::vector<bool> allocated(parameter_register_constraints.size(), false);
+
+  // Stack position of the first argument
+  unsigned int stack_offset = 16;
+
+  for (auto &argument : function.args()) {
+    anvill::ParameterDecl declaration = {};
+    declaration.type = argument.getType();
+
+    // Try to allocate from a register. If a register is not available then
+    // allocate from the stack.
+    if (remill::Register *reg = TryRegisterAllocate(
+            *argument.getType(), allocated, parameter_register_constraints)) {
+      declaration.reg = reg;
+    } else {
+      remill::Register *mem_reg =
+          new remill::Register("RSP", stack_offset, 8, 0, argument.getType());
```

Comment:
This doesn't make sense. You should never allocate a new `remill::Register`, as it will leak; you should look up the stack pointer from the `Arch`. You're passing a `stack_offset` to fill in the `offset` [1] field, which makes no sense, as that stack offset is unrelated to the `State` structure. You want a `ValueDecl` that has `mem_reg` as the stack pointer, and `mem_offset` as your `stack_offset`. See here: https://github.com/lifting-bits/anvill/blob/master/include/anvill/Decl.h#L61-L62

[1] https://github.com/lifting-bits/remill/blob/master/remill/Arch/Arch.h#L53

Reply:
Fixed 63d92a6